Ext/Accessibility/src/main/java/dev/testify/accessibility/internal/AccessibilityScreenshotLifecycleObserver.kt

```diff
@@ -0,0 +1,101 @@
+import dev.testify.ScreenshotLifecycle
+import dev.testify.accessibility.exception.AccessibilityErrorsException
+import dev.testify.internal.extensions.TestInstrumentationRegistry
+import dev.testify.internal.extensions.TestInstrumentationRegistry.Companion.instrumentationPrintln
+import dev.testify.internal.extensions.TestInstrumentationRegistry.Companion.isRecordMode
+import dev.testify.internal.extensions.cyan
+import dev.testify.testDescription
+import java.util.Locale
+
+internal class AccessibilityScreenshotLifecycleObserver(private val context: Context) : ScreenshotLifecycle {
+
+    /**
+     * The screenshot is taken after the activity is fully loaded and rendered. This provides an opportune moment to run
+     * the accessibility checks.
+     *
+     * Here, we run the checks and compare the results to the baseline json.
+     */
+    override fun afterScreenshot(activity: Activity, currentBitmap: Bitmap?) {
+        super.afterScreenshot(activity, currentBitmap)
+
+        val results = runChecksAndSaveBaseline(activity)
+        if (loadBaselineAndCompare(results).not()) {
+            val description = getInstrumentation().testDescription
+            if (isRecordMode) {
+                instrumentationPrintln("\n\t✓ " + "Recording accessibility baseline for ${description.name}".cyan())
+            } else {
+                throw AccessibilityErrorsException(
+                    results,
+                    activity,
+                    description,
+                    TestInstrumentationRegistry.getModuleName()
+                )
+            }
+        }
+    }
+
+    private fun runChecksAndSaveBaseline(activity: Activity): CheckResults {
+        return runAccessibilityChecks(activity).also {
+            writeAccessibilityCheckResults(it)
+        }
+    }
+
+    private fun loadBaselineAndCompare(currentResults: CheckResults): Boolean {
+        val baselineResults = readAccessibilityCheckResults(context)
+        return (baselineResults == currentResults)
+    }
+
+    private fun runAccessibilityChecks(activity: Activity): CheckResults {
+        val view = activity.findViewById<View>(android.R.id.content)
+        val checks = getAccessibilityHierarchyChecksForPreset(LATEST)
+        val hierarchy = AccessibilityHierarchyAndroid.newBuilder(view).build()
+        val results = checks.flatMap { it.runCheckOnHierarchy(hierarchy) }
+        return CheckResults(results.filter { it.type != NOT_RUN }.map { it.toCheckResult() })
+    }
+
+    private fun AccessibilityHierarchyCheckResult.toCheckResult(): CheckResult {
+        return CheckResult(
+            id = this.resultId,
+            type = this.type.name,
+            check = this.sourceCheckClass.simpleName,
+            elementClass = this.element?.className.toString(),
+            resourceName = this.element?.resourceName.toString(),
+            message = this.getMessage(Locale.getDefault()).toString()
```

Comment:
Hmm would it make sense to use `Locale.ROOT` here? Not sure how this would behave if the device was using a non EN language?

Reply:
🤔 
I wonder if the error messages are localized at all.
I'll try with ROOT though as that could be "safer"; these should be stable for all users.